### DISCUSSION TITLE
Infinite paths in @xstate/graph

### DISCUSSION THREAD

**Discussion by pipex** [statelyai/xstate#1395] [General] [answered]
Hi,

For state machines with context, it looks like `getShortestPaths` and `getSimplePaths` may not terminate.

For instance, the following machine

```typescript
const machine = Machine<{ pings: number }>({
  initial: "ready",
  context: { pings: 0 },
  states: {
    ready: {
      on: {
        PING: {
          target: "pinging",
          actions: actions.assign({ pings: ({ pings }) => pings + 1 })
        },
        DONE: "stopped"
      }
    },
    pinging: {
      on: {
        DONE: "ready"
      }
    },
    stopped: {
      type: "final"
    }
  }
});
```

Since there is a loop between the states `pinging` and `ready`, I get the error `too much recursion`

Is this intended behavior? Any advice on how to prevent this?

Here is a link with an example sandbox 
https://codesandbox.io/s/compassionate-worker-iuw01?file=/src/index.ts

**Comment by davidkpiano** [comment, marked as answer]
Yes, it is intended behavior. Finite states are finite; extended state (in `context`) can be infinite.

The overall state is a combination of finite state (`ready` or `pinging` or `stopped`) and extended state (`count: 0 ... ∞`)

You have to limit the state:

```js
getShortestPaths(machine, {
  filter: (state) => state.context.pings < 10
})
```

**Reply by pipex** [reply]
Oh, that actually makes a lot of sense now that you put it that way. Thanks for the answer!

> Finite states are finite; extended state (in context) can be infinite.
